### DISCUSSION TITLE
The possibility of Fornjot as an alternative CAD kernel for CadQuery

### DISCUSSION THREAD

**Discussion by huskier** [hannobraun/fornjot#2317] [General]
[CadQuery](https://github.com/CadQuery/cadquery) is an intuitive, easy-to-use Python module for building parametric 3D CAD models. At present, the underlying CAD kernel for CadQuery is OpenCascade. Multiple CAD kernels supporting was considered in the initial design of CadQuery. However, the open source CAD kernel is rare. To experiment with adaptation to other CAD kernels, I've created a [minimal CadQuery](https://github.com/huskier/MinimalCadQuery). IMO, another kernel may benefit the development of CadQuery.

To make the adaption, we firstly need to create a Python binding for Fornjot. I wonder whether somebody has done the Python binding thing? Through the web search, I have found no work on the binding.

On top of the above mentioned Python binding, Fornjot should support the following CAD types:
Vertex
Edge
Wire
Face
Shell
Solid
CompSolid
Compound

A STEP exporting function should be there for output.

@hannobraun I'd like to see that the CadQuery could work with Fornjot, and I'd like to make this happen, even if only a simplest prototype at the first beginning.

**Comment by hannobraun** [comment]
Hey @huskier, thank you for your interest in Fornjot!

I think using Fornjot as a kernel for CadQuery is an exciting possibility, but I worry that the project is not ready for that. Specifically, the geometry that Fornjot can represent is very limited, which is something I'm working on fixing.

> To make the adaption, we firstly need to create a Python binding for Fornjot. I wonder whether somebody has done the Python binding thing? Through the web search, I have found no work on the binding.

To my knowledge, no Python binding exists for Fornjot.


> On top of the above mentioned Python binding, Fornjot should support the following CAD types:
> Vertex
> Edge
> Wire
> Face
> Shell
> Solid
> CompSolid
> Compound

Fornjot supports most of those. It doesn't represent wires, and I don't know what `CompSolid` and `Compound` are.


> A STEP exporting function should be there for output.

This is not supported. And I don't think it makes sense to work on that until https://github.com/hannobraun/fornjot/issues/2118 is done.

---

Since this doesn't track actionable work that needs to be done within this repository (except for STEP support, which is on the [feature wishlist](https://github.com/hannobraun/fornjot/discussions/146), I'm converting this to a discussion.

**Reply by huskier** [reply]
Thank you for your explanation. I will keep tracing the status of the Fornjog development. If I have found something I can do, I would like to jump into the work.

**Comment by jdegenstein** [comment]
I am a contributor on the [build123d](https://build123d.readthedocs.io/en/latest/) project which is related to CadQuery and also depends on OCCT. I have been following Fornjot for awhile and I am also interested in the eventual potential for CQ+Fornjot and build123d+Fornjot.

Regarding bindings, it might be worthwhile to investigate projects like PyO3. Given hannobraun's response, I don't know if this effort is worthwhile yet. I definitely would not want to put an additional drain on the resources here.

**Reply by hannobraun** [reply]
Thank you for the info, @jdegenstein!

> Given hannobraun's response, I don't know if this effort is worthwhile yet.

It might be a worthwhile project to start experimenting with Python bindings, and provide feedback. I have absolutely no idea how PyO3 works, for example, and for all I know I might do something easily avoided, that prevents PyO3 from working for this project.

I do definitely think it is too early to create Python bindings with the expectation of having anything that is useful for CadQuery or build123d. Fornjot's functionality is simply too basic for that, and that won't even start to change until https://github.com/hannobraun/fornjot/issues/2118 (or something like it) is implemented and stable enough to be built on top of.


> I definitely would not want to put an additional drain on the resources here.

To be very clear, any Python bindings would be a 3rd-party effort and not be maintained by me. There's simply no way I can do it, given current levels of funding.

**Reply by kcleung** [reply]
> I am a contributor on the [build123d](https://build123d.readthedocs.io/en/latest/) project which is related to CadQuery and also depends on OCCT. I have been following Fornjot for awhile and I am also interested in the eventual potential for CQ+Fornjot and build123d+Fornjot.
> 
> Regarding bindings, it might be worthwhile to investigate projects like PyO3. Given hannobraun's response, I don't know if this effort is worthwhile yet. I definitely would not want to put an additional drain on the resources here.

Alternatively, we should aim at porting build123d itself to fornjot and/or truck as a long-term goal.  This will involving implementing all of the following CAD types:

Vertex
Edge
Wire
Face
Shell
Solid
CompSolid
Part
Compound / Assembly
Joints (which is used to specify the joint between the parts)

The original build123d is a library layer above the ocp and opencascade layer.

However, in this case, what I propose is to implement the build123d features and CAD types into the fornjot/truck layer itself.

**Reply by jdegenstein** [reply]
> I definitely would not want to put an additional drain on the resources here.

@kcleung I stand by this statement. OpenCASCADE has over 1 million lines of code and the development started sometime around 1980 with many people working on it. Putting build123d into fornjot is a big job unto itself -- let alone the discussion about project maturity.

**Reply by hannobraun** [reply]
Thank you for your comments, @kcleung and @jdegenstein.

I'm not very familiar with build123d or CadQuery, and I don't know what specifically they provide on top of OpenCASCADE (in addition to a nice Python API). So when we're talking about integrating this kind of functionality into Fornjot versus building it on top, I don't really know what that means.

But I can say a few things in general:

- The way Fornjot is built, I don't think that code inside of the libraries here has any special privilege versus code built on top. So whatever work is required for such a project, it should be possible to do outside of this repository, by a third party. And if it isn't, I consider that a bug.
- I welcome any discussion about how Fornjot could be used, but I'd like to make clear (as I have before), that I'm strapped for resources. I'm still putting quite a bit of time into this project, very regularly, but it's really the bare minimum to get _anything_ done in this space. Anything on top of what I'm already working on has to start as someone else's initiative, outside of this repository. So @jdegenstein is definitely on the right track :smile: (What we can eventually merge back and maintain here is a separate discussion, which we can have once there's actual code to have that discussion about.)
- Right now is probably the weirdest time to start building anything on top of Fornjot, as I've embarked on a [series of experiments](https://github.com/hannobraun/fornjot/tree/main/experiments) a few months ago. The final outcome is unclear, but I know for certain that the these experiments will at the very least significantly alter Fornjot's architecture, if not outright replace the current mainline code.
- That means that any work on top of Fornjot as it currently is, will become obsolete soon. Building on top of my experimental code is probably not very practical at all, if your goals is to create something beneficial for end users. I don't know what that means for an initiative such as the one proposed here. If anybody's seriously interested in working on that, I suggest getting in touch and coordinating closely with me.
